### PR TITLE
liam/ledger funding

### PR DESCRIPTION
**Pull request by snario**
**🚧 work in progress 🚧**

This pull request implements ledger funding support into the `server-wallet`. The implementation is based on top of the objectives based architecture implemented thus far in the [`liam/store-objectives`](https://github.com/statechannels/statechannels/tree/liam/store-objectives) branch and which is explained in detail in [Tom's server wallet architecture document](https://www.notion.so/elfour/Server-wallet-updated-architecture-ad0372e611ac485aa16dc2c4df65a14f). 

### 🤖 State Machine
The **OpenChannel** objective's protocol is extended to include additional (a) protocol state and (b) protocol actions to accommodate opening channels which request ledger funding as their strategy.

- (a) the protocol state is extended by adding the following:
  ```ts
  type ProtocolState = { // ...
    fundingChannel?: ChannelState;
    ledgerFundingRequested?: boolean;
    channelsPendingRequest?: ChannelState[];
  }
  ```
  These new properties are computed via a query made to the database that depends on the data included in the objective. For example, is an `OpenChannel` object has `fundingStrategy` as `"Direct"` then all of these properties are `undefined`. However, if `fundingStrategy` is `"Ledger"` then they will all have values. 

- (b) the protocol actions are extended by adding the following:
  ```ts
  type OpenChannelProtocolAction = // ...
    | RequestLedgerFunding
    | MarkLedgerFundingRequestsAsComplete;
  ```
  - `RequestLedgerFunding` tells the wallet that it would like to request a channel be funded by a ledger channel.
  - `MarkLedgerFundingRequestsAsComplete` tells the wallet to consider a request made for funding to have been handled.

The protocol additions can be roughly described like this, given a supported prefund setup of the app and the objective is expecting to fund the channel by the `"Ledger"` funding strategy.

1. Create a `LedgerRequest` if needed via `RequestLedgerFunding`
2. Check if there are any pending ledger updates
    1. If there are, iteratively allocate funds from the supported state's (i.e., **S**) outcome to each request[2], call this **O₁** — re-use **O₁** on each iteration of the algorithm (e.g., if there is a conflict below, don't reassess based on any _new_ pending ledger updates which may have occurred in the meanwhile)
    3. Check if there is a conflicting ledger update, **O₂**, from a counterparty such that **O₁ ⋂ O₂ ≠ O₁**
         1. If so sign an outcome for **O₁ ⋂ O₂** with turn number 2 higher than the one corresponding to **O₂** instead
         2. Otherwise sign **O₁** with turn number 2 higher than **S**
3. If there are no pending ledger updates, then check the outcome of **S** against the pending updates
    1. For any pending request that is in the outcome of **S** mark that request as "succeeded" via `MarkLedgerFundingRequestsAsComplete`

### 💾 Store
The **Store** is tasked with storing and managing new types of data. In particular:

- **Ledgers**: a table which maps a ledger channel's `channelId` to its `assetHolderAddress`. This is meant to represent the wallet's knowledge that a particular channel is funded on-chain. [1]

- **Ledger Requests**: a table which keeps track of requests made to allocate funds from a ledger channel to an application channel. This table is expected to be operated on to compute a new ledger channel update that two or more parties can then agree on — not all ledger requests will be satisfied. 

Additionally, the Store is also now responsible for computing protocol state for an objective.

### ✅ Testing  Strategy

**Wallet to Wallet E2E**
- [x] Allows two `Wallet`s to create a ledger channel and an application channel and open the application channel via the ledger funding strategy and get it to the "running" status

**Wallet API**
- Joining a channel that has proposed funding via ledger
  - [x] Signs the pre-fund setup and produces a ledger update proposal
- Pushing messages into the wallet
  - [x] Produces a ledger update proposal if it receives a now-supported application pre-fund setup
  - [x] Countersigns a ledger update proposal and countersigns pre-fund setup if provided at same time
  - [x] Countersigns a ledger update proposal and produces a post-fund setup
  - [x] Negotiates a ledger update if it receives a conflicting proposal 

### 🤔 Open Questions

- [1] How should a wallet decide a particular channel is designated as a ledger channel? Right now there is a method on `Wallet` called `__setLedger` which is just for the time being.

- [2] What should the order be of requests that get funded? `channelNonce`?

### CODE REVIEW

File: packages/server-wallet/src/protocols/ledger-funding.ts [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+  channelsPendingRequest,
+}: ProtocolState): SignLedgerStateForRequests | false => {
+  if (!supported) return false;
+  if (!latestSignedByMe) return false;
+
+  const supportedOutcome = checkThat(supported.outcome, isSimpleAllocation);
+  const latestOutcome = checkThat(latest.outcome, isSimpleAllocation);
+  const myLatestOutcome = checkThat(latestSignedByMe.outcome, isSimpleAllocation);
+
+  const counterPartyProposedNewUpdate = latest.turnNum === latestSignedByMe.turnNum + 2;
```
Comment (geoknee) [inline packages/server-wallet/src/protocols/ledger-funding.ts:70]: Is this assuming a 2 party channel?

**Comment by snario** [comment]
Added a wallet-to-wallet test. (cc @geoknee)

image

File: packages/server-wallet/src/protocols/ledger-funding.ts (hunk 1 above)
Comment (snario) [reply packages/server-wallet/src/protocols/ledger-funding.ts]: Yeah, it could be `+ n` I suppose... need to think more about how that works with CRDTs in this case.

File: packages/server-wallet/src/utilities/messaging.ts [hunk 2]
```diff
@@ -44,8 +44,8 @@ function mergeProp<T>(a: T[] | undefined, b: T[] | undefined): T[] | undefined {
   if (a && b) {
     return _.orderBy(
       _.uniqWith(_.concat(a, b), _.isEqual),
-      ['channelId', 'turnNum'],
-      ['desc', 'asc']
+      ['channelNonce', 'turnNum'],
+      ['asc', 'asc']
```
Comment (snario) [inline packages/server-wallet/src/utilities/messaging.ts:48]: Flagging as a possible separate PR. To discuss: what order should `channelResults` and `signedStates` be in? I dislike using alphanumerically sorted `channelId`s since they are harder to predict.

File: packages/server-wallet/src/utilities/messaging.ts [hunk 3]
```diff
@@ -57,7 +57,7 @@ function mergeProp<T>(a: T[] | undefined, b: T[] | undefined): T[] | undefined {
 }
 
 export function mergeChannelResults(channelResults: ChannelResult[]): ChannelResult[] {
-  const sorted = _.orderBy(_.reverse(channelResults), ['channelId', 'turnNum'], ['desc', 'desc']);
+  const sorted = _.orderBy(channelResults, ['channelNonce', 'turnNum'], ['asc', 'desc']);
```
Comment (snario) [inline packages/server-wallet/src/utilities/messaging.ts:60]: Same as https://github.com/statechannels/statechannels/pull/2699/files#r506781828

File: packages/server-wallet/src/wallet/__test__/integration/join-channel.test.ts [hunk 4]
```diff
@@ -1,13 +1,22 @@
-import {simpleEthAllocation, serializeOutcome, BN} from '@statechannels/wallet-core';
+import {
+  simpleEthAllocation,
+  serializeOutcome,
+  BN,
+  serializeState,
+  makeDestination,
+  SignedStateWithHash,
+} from '@statechannels/wallet-core';
+import {ETH_ASSET_HOLDER_ADDRESS} from '@statechannels/wallet-core/lib/src/config';
```
Comment (snario) [inline packages/server-wallet/src/wallet/__test__/integration/join-channel.test.ts:9]: 🚩 flagging to fix

File: packages/server-wallet/src/wallet/__test__/integration/join-channel.test.ts [hunk 5]
```diff
@@ -142,17 +155,15 @@ describe('directly funded app', () => {
           },
         },
       ],
-      // TODO: channelResults is not calculated correctly: see the Channel model's channelResult
-      // implementation
-      // channelResults: [{channelId, turnNum: 3, outcome, status: 'funding'}],
+      channelResult: {channelId, turnNum: 0, status: 'opening'},
```
Comment (snario) [inline packages/server-wallet/src/wallet/__test__/integration/join-channel.test.ts:158]: Could be a separate PR.

File: packages/server-wallet/src/wallet/__test__/integration/join-channel.test.ts [hunk 6]
```diff
@@ -194,3 +206,129 @@ describe('directly funded app', () => {
...
+                serializeState(stateWithHashSignedBy(alice())(expectedUpdatedLedgerState)),
+              ],
+            },
+          },
+        },
+      ],
+      channelResult: {
+        channelId,
+        turnNum: 0,
+        // outcome: serializeOutcome(preFS.outcome), // TODO: (Liam) This is empty... not sure why
```
Comment (snario) [inline packages/server-wallet/src/wallet/__test__/integration/join-channel.test.ts:319]: 🚩 flagging to fix or open new issue for

File: packages/server-wallet/src/wallet/store.ts [hunk 7]
```diff
@@ -461,6 +466,124 @@ export class Store {
     return result;
   }
 
+  // FIXME: Turn into SQL tables
+  public eraseLedgerDataFromMemory(): void {
+    this.ledgerRequests = new LedgerRequests();
+    this.ledgers = {};
+  }
```
Comment (snario) [inline packages/server-wallet/src/wallet/store.ts:473]: Relevant for #2693 perhaps.  Will also need to replicate work done here to "SQL-ify" these in memory ledger constructs I have made.

File: packages/server-wallet/src/__test__/create-and-close-channel/ledger-funding-simplified.seqdiag [hunk 8]
```diff
@@ -0,0 +1,14 @@
...
+AppA->WalletA: createChannel
+WalletA-->WalletB: A-0a
+WalletB->AppB: 'proposed'
+AppB->WalletB: joinChannel
+WalletB-->WalletA: A-0b, L-5b
+WalletA->AppA: 'opening'
+WalletA-->WalletB: A-3a, L-5a
+WalletB->AppB: 'running'
+WalletB-->WalletA: A-3b
+WalletA->AppA: 'running'
```
Comment (snario) [inline packages/server-wallet/src/__test__/create-and-close-channel/ledger-funding-simplified.seqdiag:14]: cc @geoknee 

image

**Comment by snario** [comment]
Implements [the updates server architecture](https://www.notion.so/elfour/Server-wallet-updated-architecture-ad0372e611ac485aa16dc2c4df65a14f) as of 07f0e64

**Comment by snario** [comment]
Replacing with #2715
